File: bharat-fm/sdks/go/internal/domains/education.go
package domains

import (
    "context"
    "fmt"

    "github.com/bharat-ai/bharat-fm/sdks/go/internal/client"
    "github.com/bharat-ai/bharat-fm/sdks/go/internal/models"
)

// EducationAIClient provides education-specific AI capabilities
type EducationAIClient struct {
    client *client.Client
}

// NewEducationAIClient creates a new education AI client
func NewEducationAIClient(client *client.Client) *EducationAIClient {
    return &EducationAIClient{
        client: client,
    }
}

// TutoringRequest represents a tutoring session request
type TutoringRequest struct {
    Subject          string   `json:"subject"`
    Topic            string   `json:"topic"`
    StudentLevel     string   `json:"student_level"`
    LearningStyle    string   `json:"learning_style,omitempty"`
    Language         string   `json:"language,omitempty"`
    PreviousKnowledge []string `json:"previous_knowledge,omitempty"`
    LearningObjectives []string `json:"learning_objectives,omitempty"`
}

// TutoringResponse represents a tutoring session response
type TutoringResponse struct {
    SessionID        string `json:"session_id"`
    TutorResponse    string `json:"tutor_response"`
    KeyConcepts      []string `json:"key_concepts"`
    LearningResources []struct {
        Type        string `json:"type"`
        Title       string `json:"title"`
        URL         string `json:"url,omitempty"`
        Description string `json:"description"`
    } `json:"learning_resources"`
    PracticeQuestions []struct {
        Question    string `json:"question"`
        Type        string `json:"type"`
        Difficulty  string `json:"difficulty"`
        Answer      string `json:"answer,omitempty"`
        Explanation string `json:"explanation,omitempty"`
    } `json:"practice_questions"`
    ProgressAssessment struct {
        Understanding      float64   `json:"understanding"`
        Confidence         float64   `json:"confidence"`
        AreasForImprovement []string `json:"areas_for_improvement"`
    } `json:"progress_assessment"`
    SessionTime float64 `json:"session_time"`
}

// ContentGenerationRequest represents a content generation request
type ContentGenerationRequest struct {
    Subject           string                 `json:"subject"`
    Topic             string                 `json:"topic"`
    ContentType       string                 `json:"content_type"`
    TargetAudience    map[string]interface{} `json:"target_audience"`
    LearningObjectives []string               `json:"learning_objectives,omitempty"`
    Duration          int                    `json:"duration,omitempty"`
    Language          string                 `json:"language,omitempty"`
    IncludeVisualAids  bool                   `json:"include_visual_aids,omitempty"`
}

// ContentGenerationResponse represents a content generation response
type ContentGenerationResponse struct {
    Content               string `json:"content"`
    Structure             []struct {
        Type     string `json:"type"`
        Title    string `json:"title,omitempty"`
        Content  string `json:"content"`
        Order    int    `json:"order"`
    } `json:"structure"`
    LearningOutcomes     []string `json:"learning_outcomes"`
    AssessmentCriteria  []string `json:"assessment_criteria,omitempty"`
    EstimatedCompletionTime float64 `json:"estimated_completion_time"`
    GenerationTime       float64 `json:"generation_time"`
}

// StartTutoringSession starts a tutoring session
func (c *EducationAIClient) StartTutoringSession(ctx context.Context, req *TutoringRequest) (*TutoringResponse, error) {
    var result TutoringResponse
    
    apiReq := map[string]interface{}{
        "subject":       req.Subject,
        "topic":         req.Topic,
        "student_level": req.StudentLevel,
    }
    
    if req.LearningStyle != "" {
        apiReq["learning_style"] = req.LearningStyle
    }
    if req.Language != "" {
        apiReq["language"] = req.Language
    }
    if len(req.PreviousKnowledge) > 0 {
        apiReq["previous_knowledge"] = req.PreviousKnowledge
    }
    if len(req.LearningObjectives) > 0 {
        apiReq["learning_objectives"] = req.LearningObjectives
    }

    resp, err := c.client.GetHttpClient().R().
        SetContext(ctx).
        SetBody(apiReq).
        SetResult(&result).
        Post("/education/tutoring")
    
    if err != nil {
        return nil, fmt.Errorf("tutoring session request failed: %w", err)
    }
    
    if resp.StatusCode() != 200 {
        return nil, fmt.Errorf("tutoring session API error: status %d", resp.StatusCode())
    }
    
    return &result, nil
}

// GenerateContent generates educational content
func (c *EducationAIClient) GenerateContent(ctx context.Context, req *ContentGenerationRequest) (*ContentGenerationResponse, error) {
    var result ContentGenerationResponse
    
    apiReq := map[string]interface{}{
        "subject":         req.Subject,
        "topic":           req.Topic,
        "content_type":    req.ContentType,
        "target_audience": req.TargetAudience,
    }
    
    if len(req.LearningObjectives) > 0 {
        apiReq["learning_objectives"] = req.LearningObjectives
    }
    if req.Duration > 0 {
        apiReq["duration"] = req.Duration
    }
    if req.Language != "" {
        apiReq["language"] = req.Language
    }
    apiReq["include_visual_aids"] = req.IncludeVisualAids

    resp, err := c.client.GetHttpClient().R().
        SetContext(ctx).
        SetBody(apiReq).
        SetResult(&result).
        Post("/education/generate-content")
    
    if err != nil {
        return nil, fmt.Errorf("content generation request failed: %w", err)
    }
    
    if resp.StatusCode() != 200 {
        return nil, fmt.Errorf("content generation API error: status %d", resp.StatusCode())
    }
    
    return &result, nil
}

// GetSubjectCurriculum gets subject curriculum
func (c *EducationAIClient) GetSubjectCurriculum(ctx context.Context, subject, level string) (*struct {
    Subject      string `json:"subject"`
    Level        string `json:"level"`
    Units        []struct {
        Name               string   `json:"name"`
        Topics             []string `json:"topics"`
        LearningObjectives []string `json:"learning_objectives"`
        EstimatedHours     int      `json:"estimated_hours"`
        Prerequisites      []string `json:"prerequisites,omitempty"`
    } `json:"units"`
    TotalHours   int      `json:"total_hours"`
    Skills       []string `json:"skills"`
}, error) {
    var result struct {
        Subject      string `json:"subject"`
        Level        string `json:"level"`
        Units        []struct {
            Name               string   `json:"name"`
            Topics             []string `json:"topics"`
            LearningObjectives []string `json:"learning_objectives"`
            EstimatedHours     int      `json:"estimated_hours"`
            Prerequisites      []string `json:"prerequisites,omitempty"`
        } `json:"units"`
        TotalHours   int      `json:"total_hours"`
        Skills       []string `json:"skills"`
    }

    resp, err := c.client.GetHttpClient().R().
        SetContext(ctx).
        SetResult(&result).
        Get(fmt.Sprintf("/education/curriculum/%s/%s", subject, level))
    
    if err != nil {
        return nil, fmt.Errorf("get curriculum request failed: %w", err)
    }
    
    if resp.StatusCode() != 200 {
        return nil, fmt.Errorf("get curriculum API error: status %d", resp.StatusCode())
    }
    
    return &result, nil
}

// SimpleStartTutoringSession provides a simple tutoring session interface
func (c *EducationAIClient) SimpleStartTutoringSession(ctx context.Context, subject, topic, studentLevel string) (string, error) {
    req := &TutoringRequest{
        Subject:      subject,
        Topic:        topic,
        StudentLevel: studentLevel,
    }
    
    resp, err := c.StartTutoringSession(ctx, req)
    if err != nil {
        return "", err
    }
    
    return resp.TutorResponse, nil
}

// SimpleGenerateContent provides a simple content generation interface
func (c *EducationAIClient) SimpleGenerateContent(ctx context.Context, subject, topic, contentType string) (string, error) {
    targetAudience := map[string]interface{}{
        "level": "secondary",
        "age":   14,
    }
    
    req := &ContentGenerationRequest{
        Subject:        subject,
        Topic:          topic,
        ContentType:    contentType,
        TargetAudience: targetAudience,
    }
    
    resp, err := c.GenerateContent(ctx, req)
    if err != nil {
        return "", err
    }
    
    return resp.Content, nil
}

File: bharat-fm/sdks/javascript/src/domains/education.ts
/**
 * Education AI domain-specific client for BFMF
 */

import { BharatAIClient } from '../client';
import { GenerationRequest, GenerationResponse } from '../types';
import { BharatAIError } from '../errors';

/**
 * Tutoring session request
 */
export interface TutoringRequest {
  /** Subject */
  subject: string;
  /** Topic/Concept */
  topic: string;
  /** Student level */
  studentLevel: 'primary' | 'secondary' | 'higher-secondary' | 'undergraduate' | 'postgraduate';
  /** Learning style */
  learningStyle?: 'visual' | 'auditory' | 'kinesthetic' | 'reading';
  /** Language preference */
  language?: string;
  /** Previous knowledge */
  previousKnowledge?: string[];
  /** Learning objectives */
  learningObjectives?: string[];
}

/**
 * Tutoring session response
 */
export interface TutoringResponse {
  /** Session ID */
  sessionId: string;
  /** Tutor response */
  tutorResponse: string;
  /** Key concepts covered */
  keyConcepts: string[];
  /** Learning resources */
  learningResources: Array<{
    type: 'video' | 'article' | 'exercise' | 'simulation';
    title: string;
    url?: string;
    description: string;
  }>;
  /** Practice questions */
  practiceQuestions: Array<{
    question: string;
    type: 'multiple-choice' | 'short-answer' | 'essay' | 'problem-solving';
    difficulty: 'easy' | 'medium' | 'hard';
    answer?: string;
    explanation?: string;
  }>;
  /** Progress assessment */
  progressAssessment: {
    understanding: number; // 0-100
    confidence: number; // 0-100
    areasForImprovement: string[];
  };
  /** Session time */
  sessionTime: number;
}

/**
 * Content generation request
 */
export interface ContentGenerationRequest {
  /** Subject */
  subject: string;
  /** Topic */
  topic: string;
  /** Content type */
  contentType: 'lesson-plan' | 'study-material' | 'assignment' | 'quiz' | 'presentation';
  ** Target audience */
  targetAudience: {
    level: 'primary' | 'secondary' | 'higher-secondary' | 'undergraduate' | 'postgraduate';
    age?: number;
    background?: string;
  };
  ** Learning objectives */
  learningObjectives?: string[];
  ** Duration (in minutes) */
  duration?: number;
  ** Language */
  language?: string;
  ** Include visual aids */
  includeVisualAids?: boolean;
}

/**
 * Content generation response
 */
export interface ContentGenerationResponse {
  /** Generated content */
  content: string;
  /** Content structure */
  structure: Array<{
    type: 'heading' | 'paragraph' | 'list' | 'table' | 'image' | 'video' | 'exercise';
    title?: string;
    content: string;
    order: number;
  }>;
  /** Learning outcomes */
  learningOutcomes: string[];
  /** Assessment criteria */
  assessmentCriteria?: string[];
  /** Estimated completion time */
  estimatedCompletionTime: number;
  /** Generation time */
  generationTime: number;
}

/**
 * Assessment generation request
 */
export interface AssessmentGenerationRequest {
  /** Subject */
  subject: string;
  /** Topic */
  topic: string;
  /** Assessment type */
  assessmentType: 'quiz' | 'assignment' | 'exam' | 'project';
  /** Difficulty level */
  difficulty: 'easy' | 'medium' | 'hard' | 'mixed';
  /** Number of questions */
  numberOfQuestions: number;
  /** Question types */
  questionTypes?: ('multiple-choice' | 'short-answer' | 'essay' | 'problem-solving' | 'true-false')[];
  /** Student level */
  studentLevel: 'primary' | 'secondary' | 'higher-secondary' | 'undergraduate' | 'postgraduate';
  ** Time limit (in minutes) */
  timeLimit?: number;
  ** Total marks */
  totalMarks?: number;
  ** Language */
  language?: string;
}

/**
 * Assessment generation response
 */
export interface AssessmentGenerationResponse {
  /** Assessment ID */
  assessmentId: string;
  /** Questions */
  questions: Array<{
    id: string;
    question: string;
    type: 'multiple-choice' | 'short-answer' | 'essay' | 'problem-solving' | 'true-false';
    difficulty: 'easy' | 'medium' | 'hard';
    marks: number;
    options?: string[]; // For multiple choice
    correctAnswer?: string;
    explanation?: string;
  }>;
  /** Answer key */
  answerKey: Record<string, string>;
  /** Rubric */
  rubric?: {
    criteria: Array<{
      name: string;
      description: string;
      maxMarks: number;
    }>;
  };
  /** Instructions */
  instructions: string;
  /** Estimated completion time */
  estimatedCompletionTime: number;
  /** Generation time */
  generationTime: number;
}

/**
 * Progress tracking request
 */
export interface ProgressTrackingRequest {
  /** Student ID */
  studentId: string;
  ** Subject */
  subject: string;
  ** Topics covered */
  topicsCovered: string[];
  ** Assessment scores */
  assessmentScores: Array<{
    assessmentId: string;
    score: number;
    maxScore: number;
    date: string;
  }>;
  ** Learning goals */
  learningGoals?: string[];
  ** Time spent (in minutes) */
  timeSpent?: number;
}

/**
 * Progress tracking response
 */
export interface ProgressTrackingResponse {
  /** Student ID */
  studentId: string;
  ** Overall progress */
  overallProgress: {
    percentage: number;
    grade: 'A+' | 'A' | 'B' | 'C' | 'D' | 'F';
    status: 'excellent' | 'good' | 'average' | 'needs-improvement';
  };
  ** Topic-wise progress */
  topicProgress: Array<{
    topic: string;
    mastery: number; // 0-100
    status: 'mastered' | 'in-progress' | 'needs-attention';
    lastStudied: string;
    timeSpent: number;
  }>;
  ** Strengths */
  strengths: string[];
  ** Areas for improvement */
  areasForImprovement: string[];
  ** Recommendations */
  recommendations: Array<{
    type: 'study-material' | 'practice' | 'tutoring' | 'assessment';
    title: string;
    description: string;
    priority: 'high' | 'medium' | 'low';
  }>;
  ** Predicted performance */
  predictedPerformance: {
    nextAssessment: number; // Predicted score
    confidence: number; // 0-100
  };
  /** Analysis time */
  analysisTime: number;
}

/**
 * Education AI domain-specific client
 */
export class EducationAIClient {
  private client: BharatAIClient;
  private domainEndpoint: string;

  /**
   * Create a new Education AI client
   */
  constructor(client: BharatAIClient, domainEndpoint: string = '/education') {
    this.client = client;
    this.domainEndpoint = domainEndpoint;
  }

  /**
   * Start tutoring session
   */
  async startTutoringSession(request: TutoringRequest): Promise<TutoringResponse> {
    const response = await this.client['makeRequest']<TutoringResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/tutoring`,
      data: {
        subject: request.subject,
        topic: request.topic,
        student_level: request.studentLevel,
        learning_style: request.learningStyle,
        language: request.language,
        previous_knowledge: request.previousKnowledge,
        learning_objectives: request.learningObjectives
      }
    });

    return response;
  }

  /**
   * Continue tutoring session
   */
  async continueTutoringSession(sessionId: string, userMessage: string): Promise<TutoringResponse> {
    const response = await this.client['makeRequest']<TutoringResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/tutoring/${sessionId}/continue`,
      data: {
        user_message: userMessage
      }
    });

    return response;
  }

  /**
   * Generate educational content
   */
  async generateContent(request: ContentGenerationRequest): Promise<ContentGenerationResponse> {
    const response = await this.client['makeRequest']<ContentGenerationResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/generate-content`,
      data: {
        subject: request.subject,
        topic: request.topic,
        content_type: request.contentType,
        target_audience: request.targetAudience,
        learning_objectives: request.learningObjectives,
        duration: request.duration,
        language: request.language,
        include_visual_aids: request.includeVisualAids
      }
    });

    return response;
  }

  /**
   * Generate assessment
   */
  async generateAssessment(request: AssessmentGenerationRequest): Promise<AssessmentGenerationResponse> {
    const response = await this.client['makeRequest']<AssessmentGenerationResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/generate-assessment`,
      data: {
        subject: request.subject,
        topic: request.topic,
        assessment_type: request.assessmentType,
        difficulty: request.difficulty,
        number_of_questions: request.numberOfQuestions,
        question_types: request.questionTypes,
        student_level: request.studentLevel,
        time_limit: request.timeLimit,
        total_marks: request.totalMarks,
        language: request.language
      }
    });

    return response;
  }

  /**
   * Track student progress
   */
  async trackProgress(request: ProgressTrackingRequest): Promise<ProgressTrackingResponse> {
    const response = await this.client['makeRequest']<ProgressTrackingResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/track-progress`,
      data: {
        student_id: request.studentId,
        subject: request.subject,
        topics_covered: request.topicsCovered,
        assessment_scores: request.assessmentScores,
        learning_goals: request.learningGoals,
        time_spent: request.timeSpent
      }
    });

    return response;
  }

  /**
   * Get personalized learning path
   */
  async getPersonalizedLearningPath(params: {
    studentId: string;
    subject: string;
    currentLevel: string;
    goals: string[];
    availableTime: number; // hours per week
    preferredLanguage?: string;
  }): Promise<{
    pathId: string;
    weeks: Array<{
      week: number;
      topics: Array<{
        topic: string;
        estimatedHours: number;
        resources: Array<{
          type: 'video' | 'reading' | 'exercise' | 'project';
          title: string;
          url?: string;
          description: string;
        }>;
        assessments: Array<{
          type: 'quiz' | 'assignment' | 'project';
          title: string;
          estimatedTime: number;
        }>;
      }>;
      milestones: string[];
    }>;
    totalEstimatedHours: number;
    completionDate: string;
  }> {
    const response = await this.client['makeRequest']<{
      pathId: string;
      weeks: Array<{
        week: number;
        topics: Array<{
          topic: string;
          estimatedHours: number;
          resources: Array<{
            type: 'video' | 'reading' | 'exercise' | 'project';
            title: string;
            url?: string;
            description: string;
          }>;
          assessments: Array<{
            type: 'quiz' | 'assignment' | 'project';
            title: string;
            estimatedTime: number;
          }>;
        }>;
        milestones: string[];
      }>;
      totalEstimatedHours: number;
      completionDate: string;
    }>({
      method: 'POST',
      url: `${this.domainEndpoint}/learning-path`,
      data: params
    });

    return response;
  }

  /**
   * Grade assessment automatically
   */
  async autoGradeAssessment(assessmentId: string, submissions: Array<{
    studentId: string;
    answers: Record<string, string>;
    submissionTime: string;
  }>): Promise<Array<{
    studentId: string;
    score: number;
    maxScore: number;
    percentage: number;
    grade: string;
    feedback: string;
    questionFeedback: Record<string, {
      score: number;
      feedback: string;
      correct: boolean;
    }>;
    timeSpent: number;
  }>> {
    const response = await this.client['makeRequest']<Array<{
      studentId: string;
      score: number;
      maxScore: number;
      percentage: number;
      grade: string;
      feedback: string;
      questionFeedback: Record<string, {
        score: number;
        feedback: string;
        correct: boolean;
      }>;
      timeSpent: number;
    }>>({
      method: 'POST',
      url: `${this.domainEndpoint}/grade-assessment/${assessmentId}`,
      data: { submissions }
    });

    return response;
  }

  /**
   * Get subject curriculum
   */
  async getSubjectCurriculum(subject: string, level: string): Promise<{
    subject: string;
    level: string;
    units: Array<{
      name: string;
      topics: string[];
      learningObjectives: string[];
      estimatedHours: number;
      prerequisites?: string[];
    }>;
    totalHours: number;
    skills: string[];
  }> {
    const response = await this.client['makeRequest']<{
      subject: string;
      level: string;
      units: Array<{
        name: string;
        topics: string[];
        learningObjectives: string[];
        estimatedHours: number;
        prerequisites?: string[];
      }>;
      totalHours: number;
      skills: string[];
    }>({
      method: 'GET',
      url: `${this.domainEndpoint}/curriculum/${subject}/${level}`
    });

    return response;
  }

  /**
   * Generate interactive exercise
   */
  async generateInteractiveExercise(params: {
    subject: string;
    topic: string;
    exerciseType: 'simulation' | 'game' | 'lab' | 'case-study';
    difficulty: 'beginner' | 'intermediate' | 'advanced';
    duration?: number;
    learningObjectives?: string[];
  }): Promise<{
    exerciseId: string;
    title: string;
    description: string;
    instructions: string;
    interactiveElements: Array<{
      type: 'drag-drop' | 'multiple-choice' | 'input' | 'simulation';
      question: string;
      options?: string[];
      correctAnswer?: string;
      explanation?: string;
    }>;
    learningOutcomes: string[];
    estimatedTime: number;
  }> {
    const response = await this.client['makeRequest']<{
      exerciseId: string;
      title: string;
      description: string;
      instructions: string;
      interactiveElements: Array<{
        type: 'drag-drop' | 'multiple-choice' | 'input' | 'simulation';
        question: string;
        options?: string[];
        correctAnswer?: string;
        explanation?: string;
      }>;
      learningOutcomes: string[];
      estimatedTime: number;
    }>({
      method: 'POST',
      url: `${this.domainEndpoint}/interactive-exercise`,
      data: params
    });

    return response;
  }
}